astr303a.py

```diff
@@ -0,0 +1,571 @@
+    ellipseA = ((1.5*180)/(np.pi*dcoma))*ratio #set major axis
+    ellipse = Ellipse((comaRA,comaDEC),width=2*(ellipseA),height=2*(1.5*180)/(dcoma*np.pi),linestyle=':',fill=False)
+    fig = plt.gcf()
+    ax = fig.gca()
+    plt.legend(loc='upper center',ncol=3,fontsize=10,numpoints=1)
+    plt.gca().invert_xaxis()
+    ax.add_patch(ellipse)
+    plt.grid()
+    pp.savefig()
+    plt.clf()
+
+    #create list of all redshifts
+    redshift = []
+    redshift2 = []
+    for i in range(0,len(dataarray)):
+        redshift.append(dataarray[i][6])
+        redshift2.append(dataarray[i][6])
+    for i in range(0,len(dataarray2)):
+        redshift2.append(dataarray2[i][6])
+
+    #Gather data in bins of width 0.001. Retrive counts from each bin.
+    histogram = py.hist(redshift,bins=np.arange(0.005, 0.05, 0.001),color='lightgrey')
+    counts = histogram[0] #get counts
+    shifts = histogram[1] #get z values of bins
+    binset = []
+
+    #Create list of z values at bin centers
+    for i in range(0,len(shifts)-1):
+        binset.append(shifts[i]+0.0005)
+
+    #Define Gaussian, calculate initial parameter guesses.
+    def gauss(x, a, x0, sigma):
+        return a*np.exp(-(x-x0)**2/(2*sigma**2))
+    stddev = np.std(redshift)
+    mean = np.mean(redshift)
+
+    #Fit gaussian
+    coeffs, pcov = optimize.curve_fit(gauss, binset, counts, p0 = [1, mean, stddev])       
+
+    #Plot histogram and fitted Gaussian.
+    plt.title('Figure 4: Coma Cluster Member Galaxy Redshifts')
+    plt.xlabel('Redshift '+r'$z$', fontsize = 12)
+    plt.ylabel('Number of galaxies', fontsize = 12)
+    plt.xlim(0,0.05)
+    plt.plot(binset, gauss(binset, *coeffs),color='black',linestyle='-')
+    plt.annotate(r'$\bar z$'+' ='+str('{:.4E}'.format(comameanz))+'\n'+r'$\sigma_z$'+' ='+str('{:.4E}'.format(abs(coeffs[2]))), xy=(0.70, 0.90), xycoords='axes fraction',fontsize = 10)
+    pp.savefig()
+    plt.clf()
+    
+    return
+
+
+def part1011(dataarray,dcoma,comaRA,comaDEC):
+    
+    #convert coma distance to parsecs
+    dcomaparsecs = dcoma*10**6
+    
+    for i in range(0,len(dataarray)):
+
+        #Newly compute values appended to rows of dataarray as follows...
+        #[objID(0),ra(1),dec(2),mag(3),magErr(4),ext(5),z(6),zErr(7),magabs(8),magabserr(9),lum(10),distance(11)]
+
+        #calculate absolute magnitude and error
+        mapp = (dataarray[i][3]-dataarray[i][5])
+        magabs = mapp - 5*np.log10(dcomaparsecs/10)
+        magabserr = dataarray[i][4]
+        dataarray[i].append(magabs)
+        dataarray[i].append(magabserr)
+        
+        #calculate solar lumonisties
+        lumin = 10**((4.76-magabs)/2.5)
+        dataarray[i].append(lumin)
+        
+        #calculate radial distance from centre of group
+        radians = projecteddistance(dataarray[i][1],dataarray[i][2],comaRA,comaDEC,dcoma)
+        distance = dcoma*radians
+        dataarray[i].append(distance)
+        
+    return dataarray
```

Comment:
Again, this is poor form to directly modify (in-place) a parameter to your function and then return the modified. Having done this, now you can only ever run part 10-11 after parts 1-9 and before parts 12-18. If you ever try to generate the plots in a different order, the code will break. So you now have multiple unstated dependency assumptions between functions that are by definition supposed to be independent.